toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -56,29 +56,65 @@ func main() {
 	kingpin.MustParse(app.Parse(os.Args[1:]))
 	logger.InitBestEffort(*logFile, *logLevel)
 
+	err := fetchPackages()
+	if err != nil {
+		logger.Log.Fatalf("Failed to fetch packages. Error: %s", err)
+	}
+}
+
+func fetchPackages() (err error) {
 	dependencyGraph, err := pkggraph.ReadDOTGraphFile(*inputGraph)
 	if err != nil {
-		logger.Log.Panicf("Failed to read graph to file. Error: %s", err)
+		err = fmt.Errorf("failed to read graph to file: %w", err)
+		return
 	}
 
 	toolchainPackages, err := schedulerutils.ReadReservedFilesList(*toolchainManifest)
 	if err != nil {
-		logger.Log.Fatalf("unable to read toolchain manifest file '%s': %s", *toolchainManifest, err)
+		err = fmt.Errorf("unable to read toolchain manifest file '%s': %w", *toolchainManifest, err)
+		return
 	}
 
-	if hasUnresolvedNodes(dependencyGraph) {
-		err = resolveGraphNodes(dependencyGraph, *inputSummaryFile, *outputSummaryFile, toolchainPackages, *disableUpstreamRepos, *stopOnFailure)
+	var cloner *rpmrepocloner.RpmRepoCloner
+	hasUnresolvedNodes := hasUnresolvedNodes(dependencyGraph)
+	if hasUnresolvedNodes {
+		// Create the worker environment
+		cloner, err = prepRpmCloner(*outDir, *disableUpstreamRepos)
 		if err != nil {
-			logger.Log.Panicf("Failed to resolve graph. Error: %s", err)
+			err = fmt.Errorf("failed to setup new cloner: %w", err)
+			return
+		}
+		defer cloner.Close()
+	}
+
+	if hasUnresolvedNodes {
+		err = resolveGraphNodes(dependencyGraph, *inputSummaryFile, *outputSummaryFile, toolchainPackages, cloner, *stopOnFailure)
+		if err != nil {
+			err = fmt.Errorf("failed to resolve graph: %w", err)
+			return
 		}
 	} else {
 		logger.Log.Info("No unresolved packages to cache")
 	}
 
+	// Write the graph to file (even if we are going to download delta RPMs, we want to save the graph with the resolved nodes first)
 	err = pkggraph.WriteDOTGraphFile(dependencyGraph, *outputGraph)
 	if err != nil {
-		logger.Log.Panicf("Failed to write cache graph to file. Error: %s", err)
+		err = fmt.Errorf("failed to write cache graph to file: %w", err)
+		return
 	}
+
+	// If we grabbed any RPMs, we need to convert them into a local repo
+	if hasUnresolvedNodes {
```

Comment:
>if hasUnresolvedNodes { [](http://example.com/codeflow?start=1&length=23)

Same, it looks like we can do it in one `if hasUnresolvedNodes(dependencyGraph)` and just keep the writing of the graph at the end of this function.

Reply:
Btw. I'm curious if the `ConvertDownloadedPackagesIntoRepo` method wouldn't be a better place for this log.

Reply:
Same as above, see https://github.com/microsoft/CBL-Mariner/pull/5595/commits/b70b493d084191cd261fdbe85038c033ed3537e6#diff-a5e9d5028e48733080cebf28dcf5c71da5627b7bc087ad29a037dcf51445f128

Reply:
Agreed about the log though will move it.

Reply:
Closed the other comment to keep everything in one.

I couldn't open the commit link - GitHub claims it's not there. Could you resend it or explain why we cannot keep only one `if hasUnresolvedNodes` check?